docs/compilers/Compiler Server.md

```diff
@@ -0,0 +1,85 @@
+Compiler Server
+===============
+
+As a performance optimization, the Roslyn C# and VB compilers support using a
+server process to execute compilations instead of executing the compilation
+in-process. Engineering has focused on three performance benefits from the server:
+
+1. Re-using a single process to do compilation amortizes the cost of JITing the
+   compiler code. This is usually the single largest performance benefit.
+1. Roslyn requires assembly references to be hydrated into a rich object
+   model to perform compilation. The compiler server holds a fixed-size cache of
+   metadata symbols, which can help compilation times if most compilations share
+   a sizable portion of metadata references (e.g., mscorlib).
+1. On Windows process creation itself is expensive, so dispatching compilations
+   directly from MSBuild saves creations of csc.exe.
+
+## Usage
+
+The primary way to use the compiler server is via the
+`Microsoft.Build.Tasks.CodeAnalysis` MSBuild task, which ships in both the
+`dotnet` SDK and Visual Studio. The build task contains a build client which,
+rather than executing the `csc.exe` or `vbc.exe`, creates the
+`VBCSCompiler.exe` process and dispatches the command line arguments to the
+server directly.
+
+If you're not using MSBuild, there's also a flag, `/shared` which can be
+passed directly to `csc.exe`, which causes `csc.exe` to behave as a client
+and dispatches the actual compilation to the server process. In this mode
+`csc.exe` doesn't perform any computation itself, it just dispatches the
+command line to the server process (starting one if one does not already
+exist), and then reports the results of the compilation (including any
+warnings/errors produced).
```

Comment:
We are successfully using /shared in Unity when invoking csc to optimize performance. However, there is an issue if there are multiple installs of different versions of Unity running in parallel, which may bundle different roslyn versions.

If I am reading this comment: https://github.com/dotnet/roslyn/issues/30722#issuecomment-432800076 correctly, it appears that we can pass a keyword to /shared to identify different compiler services. Is this correct? If yes, can we document that functionality as well? It could be the solution to our problem.

Reply:
@jechter That's correct. I'll document that in a follow up PR.